compiler/qsc_qasm3/src/semantic/lowerer.rs

```diff
@@ -1037,13 +1270,69 @@ impl Lowerer {
     }
 
     fn lower_quantum_decl(&mut self, stmt: &syntax::QubitDeclaration) -> semantic::StmtKind {
-        self.push_unimplemented_error_message("qubit decl stmt", stmt.span);
-        semantic::StmtKind::Err
+        // If there wasn't an explicit size, infer the size to be 1.
+        let (ty, size_and_span) = if let Some(size_expr) = &stmt.size {
+            let size_expr = self.lower_expr(size_expr);
+            let span = size_expr.span;
+            let size_expr = self.try_cast_expr_to_type(&Type::UInt(None, true), &size_expr);
+
+            if let Some(Some(semantic::LiteralKind::Int(val))) =
+                size_expr.map(|expr| expr.const_eval(&self.symbols))
+            {
+                if let Ok(size) = u32::try_from(val) {
+                    (
+                        Type::QubitArray(ArrayDimensions::One(size)),
+                        Some((size, span)),
+                    )
+                } else {
+                    let message = "quantum register size".into();
+                    self.push_semantic_error(SemanticErrorKind::ExprMustFitInU32(message, span));
+                    return semantic::StmtKind::Err;
+                }
+            } else {
+                let message = "quantum register size".into();
+                self.push_semantic_error(SemanticErrorKind::ExprMustBeConst(message, span));
+                return semantic::StmtKind::Err;
```

Comment:
I don't think we want to return stmtkind::err here. We pushed the errors and I think we want to use the unevaluated size expression in the returned stmt.